File: fix_progress.js
const fs = require('fs');

let content = fs.readFileSync('train-english-next/src/lib/services/progress.service.ts', 'utf-8');

content = content.replace(/import { Router, Response } from "express";\n/g, '');
content = content.replace(/import UserWordProgress.*?\n/g, 'import { UserWordProgress, Vocabulary } from "../db/models";\nimport dbConnect from "../db/connection";\n');
content = content.replace(/import Vocabulary.*\n/g, '');
content = content.replace(/import { authenticate.*?\n/g, '');
content = content.replace(/const router = Router\(\);\n/g, '');
content = content.replace(/router\.use\(.*?\n/g, '');

content = "export class ProgressService {\n" + content;

content = content.replace(/router\.(get|post|put|patch|delete)\("([^"]+)"(?:,\s*requireAdmin\s*as\s*any)?,\s*async\s*\([^)]*\)\s*=>\s*\{/g, (match, method, path) => {
    let name = path.replace(/\//g, "_").replace(/-/g, "_").replace(/:/g, "by_");
    if (name === "_") name = "index";
    name = name.replace(/^_/, "");
    
    return `  static async handle_${method.toUpperCase()}_${name}(req: any, res: any, userId: string) {`;
});

content = content.replace(/export default router;/g, '}');

content = content.replace(/(static async handle_[^{]+{)/g, '$1\n    await dbConnect();');

fs.writeFileSync('train-english-next/src/lib/services/progress.service.ts', content, 'utf-8');

File: fix_progress.py
import re

with open("train-english-next/src/lib/services/progress.service.ts", "r", encoding="utf-8") as f:
    content = f.read()

# Remove imports
content = re.sub(r'import { Router, Response } from "express";\n', '', content)
content = re.sub(r'import UserWordProgress.*?\n', 'import { UserWordProgress, Vocabulary } from "../db/models";\nimport dbConnect from "../db/connection";\n', content)
content = re.sub(r'import Vocabulary.*\n', '', content)
content = re.sub(r'import { authenticate.*?\n', '', content)
content = re.sub(r'const router = Router\(\);\n', '', content)
content = re.sub(r'router\.use\(.*?\n', '', content)

# Wrap in class
content = "export class ProgressService {\n" + content

# Convert router.get("/stats", async (req: AuthRequest, res: Response) => {
# to static async getStats(userId: string) {
def replace_route(match):
    method = match.group(1).upper()
    path = match.group(2)
    # create a name based on path
    name = path.replace("/", "_").replace("-", "_").replace(":", "by_")
    if name == "_": name = "index"
    name = name.strip("_")
    
    # We will just pass userId and req if needed
    return f"  static async handle_{method}_{name}(req: any, res: any, userId: string) {{"

content = re.sub(r'router\.(get|post|put|patch|delete)\("([^"]+)"(?:,\s*requireAdmin\s*as\s*any)?,\s*async\s*\([^)]*\)\s*=>\s*\{', replace_route, content)

# Remove export default router
content = re.sub(r'export default router;', '}', content)

# Add dbConnect to all methods
content = re.sub(r'(static async handle_[^{]+{)', r'\1\n    await dbConnect();', content)

with open("train-english-next/src/lib/services/progress.service.ts", "w", encoding="utf-8") as f:
    f.write(content)
